File: python3/Graphs/merging_communities.py
class UnionFind:
    def __init__(self, size: int):
        self.parent = [i for i in range(size)]
        self.size = [1] * size

    def union(self, x: int, y: int) -> None:
        rep_x, rep_y = self.find(x), self.find(y)
        if rep_x != rep_y:
            # If 'rep_x' represents a larger community, connect
            # 'rep_y 's community to it.
            if self.size[rep_x] > self.size[rep_y]:
                self.parent[rep_y] = rep_x
                self.size[rep_x] += self.size[rep_y]
                # Otherwise, connect 'rep_x's community to 'rep_y'.
            else:
                self.parent[rep_x] = rep_y
                self.size[rep_y] += self.size[rep_x]
        
    def find(self, x: int) -> int:
        if x == self.parent[x]:
            return x
        # Path compression.
        self.parent[x] = self.find(self.parent[x])
        return self.parent[x]

    def get_size(self, x: int) -> int:
        return self.size[self.find(x)]


class MergingCommunities:
    def __init__(self, n: int):
        self.uf = UnionFind(n)

    def connect(self, x: int, y: int) -> None:
        self.uf.union(x, y)

    def get_community_size(self, x: int) -> int:
        return self.uf.get_size(x)

File: kotlin/Graphs/MergingCommunities.kt
class UnionFind(size: Int) {

    private val parent = IntArray(size) { it }
    private val size = IntArray(size) { 1 }

    // With comment
    fun union(x: Int, y: Int) {
        val repX = find(x)
        val repY = find(y)
        if (repX != repY) {
            // If 'repX' represents a larger community, connect
            // 'repY 's community to it.
            if (size[repX] > size[repY]) {
                parent[repY] = repX
                size[repX] += size[repY]
            } else {
                // Otherwise, connect 'repX's community to 'repY'.
                parent[repX] = repY
                size[repY] += size[repX]
            }
        }
    }

    fun find(x: Int): Int {
        if (x == parent[x]) {
            return x
        }
        // Path compression.
        parent[x] = find(parent[x])
        return parent[x]
    }

    fun getSize(x: Int): Int {
        return size[find(x)]
    }
}

class MergingCommunities(n: Int) {

    private val uf = UnionFind(n)

    fun connect(x: Int, y: Int) {
        uf.union(x, y)
    }

    fun getCommunitySize(x: Int): Int {
        return uf.getSize(x)
    }
}
